Mage/src/main/java/mage/abilities/SpellAbility.java

```diff
@@ -210,6 +212,26 @@ public SpellAbility getSpellAbilityToResolve(Game game) {
         return this;
     }
 
+    public Card getCharacteristics(Game game) {
+        Spell spell = game.getSpell(this.getId());
+        if (spell != null) {
+            return spell;
+        }
+        return game.getCard(this.getSourceId());
+    }
+
+    public static SpellAbility getSpellAbilityFromEvent(GameEvent event, Game game) {
+        if (event.getType() != GameEvent.EventType.CAST_SPELL) {
+            return null;
+        }
+        Card card = game.getCard(event.getSourceId());
+        Optional<Ability> ability = card.getAbilities(game).get(event.getTargetId());
+        if (ability.isPresent() && ability.get() instanceof SpellAbility) {
+            return (SpellAbility) ability.get();
+        }
+        return card.getSpellAbility();
+    }
```

Comment:
Why do we need this getSpellAbilityFromEvent method? Currently it is not used.

Reply:
So, both of these new methods are intended to fix the interaction of alternate casting methods that modify characteristics like Morph and Bestow with rule modifying and cost modifying effects. If you'd like I can split getSpellAbilityFromEvent into a separate change where I actually use it to fix a bug. The original impetus behind this was https://github.com/magefree/mage/issues/5150.